### PR TITLE
Fixes for HCAL premixing

### PR DESCRIPTION
**Pull request by kpedro88**
The first two commits in this PR fix regressions that were unknowingly introduced before pre12 (which prevent premixing from running at all):
1) Premixing config for new packer
2) Packing and unpacking presamples value for QIE8 in new packer

The second two commits fix known issues specifically for QIE11 premixing:
3) Packing and unpacking QIE11 flag word (used for premixing error bits)
4) Disabling SiPM dark current noise during premixing step 1

The following plots compare the QIE11 ADC distribution with standard mixing (blue) vs. premixing (red).

After commits 1,2:
adc

After commits 1,2,3: (better agreement at high ADC where error bits are used)
adc

After commits 1,2,3,4: (better agreement at low ADC where single pe noise was previously duplicated)
adc

attn: @jmmans, @lihux25, @mdhildreth

### CODE REVIEW

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by kpedro88** [comment]
@slava77 @cvuosalo @civanch @mdhildreth please sign

**Comment by mdhildreth** [comment]
+1

**Comment by kpedro88** [comment]
@slava77 any more code review?

File: EventFilter/HcalRawToDigi/plugins/PackerHelp.h [hunk 1]
```diff
@@ -508,7 +516,7 @@ class UHTRpacker{
     HcalElectronicsId eid(readoutMap->lookup(detid));
     // loop over words in dataframe
     for(edm::DataFrame::iterator dfi=qiedf.begin() ; dfi!=qiedf.end(); ++dfi){      
-      if( dfi >= qiedf.end()-QIE11DataFrame::FLAG_WORDS ){
+      if( dfi >= qiedf.end() ){ //include flag word
```
Comment (slava77) [inline EventFilter/HcalRawToDigi/plugins/PackerHelp.h]: I see changes in the HcalNoise plots. Is this the place in this PR that drives the changes?

**Comment by slava77** [comment]
jenkins comparison for 10624 shows changes in the noise variables
maxZeros appears to have the most significant effect
wf10624_maxzeros

I recall that maxZeros had issues last year and was at some point abandoned. Is it still relevant?

**Comment by slava77** [comment]
On 10/5/16 9:37 AM, Kevin Pedro wrote:
> @slava77 <https://github.com/slava77> any more code review?
> 

nothing from visual inspection.

I'm not sure if my question of the observed differences in monitored
quantities will lead to more changes.

**Comment by kpedro88** [comment]
@slava77 I didn't expect this PR to cause any changes in standard workflows. I'm not sure why only these noise DQM plots would change...

**Comment by slava77** [comment]
So, what's causing the changes in regular mixing workflows (well, they are PU=0)?
Do the QIE8 presamples vary in MC or there something else.
Increased noise doesn't look like going in the right direction for MC, but may be I'm misinterpreting the variable.

**Comment by slava77** [comment]
On 10/5/16 9:50 AM, Kevin Pedro wrote:
> @slava77 <https://github.com/slava77> I didn't expect this PR to cause
> any changes in standard workflows. I'm not sure why only these noise DQM
> plots would change...

changes are only in 2017 workflows, just to be clear on the scope of
"standard workflow"

**Comment by kpedro88** [comment]
@slava77 actually, maybe this is a result of correctly propagating the QIE8 presamples info. When I enabled the new packer in #15889, the QIE8 presample value was not propagated. However, nothing showed up in the comparisons for that PR because the new HCAL 2017 workflows were introduced there.

@igv4321 @kencall @lihux25 @vkhristenko do any of you know if the presample value is used in the HCAL noise DQM plots?

**Comment by slava77** [comment]
lack of feedback on the observed differences is blocking signoff at this point.

**Comment by kpedro88** [comment]
We had some discussion of this in the HCAL software meeting today. I asked the relevant experts to look into these plots and see if they can understand the change. My hope/intuition is that the change is an improvement and not a regression (due to properly keeping presample info for QIE8). Hopefully we will get some input tonight/tomorrow.

**Comment by slava77** [comment]
Is presample the count of TS before the trigger sample? (it was set to 10 before, which wouldn't make much sense at least in places it is used).

It looks like the hcalNoise variations are matching changes in eraw (M0 energy)
e.g. in 10624.0:

all_pr16057-15480vsbaseline_ttbar13tev2017hcaldevwf10624p0c_hbherechitssorted_hbhereco__reco_obj_obj_eraw
the scale of the plot is not the best, but it shows a clear difference.
Downstream, the reduced hits (for HCAL in most cases selected based on noise flags) show a more clear difference
all_pr16057-15480vsbaseline_ttbar13tev2017hcaldevwf10624p0c_log10hbherechitssorted_reducedhcalrechits_hbhereco_reco_obj_obj_eraw

... I can't say clearly if noise or eraw distributions make sense, but given that presamples value is used, the old value of 10 was wrong anyways.
@kpedro88 please correct me if I missed something

**Comment by kpedro88** [comment]
@igv4321 briefly mentioned today that the presamples value is used in RECO, so if it was not set properly for QIE8, that could affect the HB RecHits. (10 was a default value used by the packer, which is fixed to take the correct value from the simulated digi in my second commit here.)

**Comment by abdoulline** [comment]
Sorry, not much input from my side tonight. Will try to understand what's 
going on tomorrow. Just a couple of comments:

In MC "presamples": N of TS in full array of CaloSamples / Digis _before_ 
the trigger one. E.g. TS=4 is SOI(sample of interest)=Trigger one in HB, so 
0,1,2,3 come as pre-samples. Set in sim.config: 
http://cmslxr.fnal.gov/source/SimCalorimetry/HcalSimProducers/python/hcalSimParameters_cfi.py#0054
(due to some historical reasons it's 5th TS = TS4 in C++ enumeration from 0)

In the data N_pre-samples come in RAW data itself (as it's set in 
on-line firmware config).

All-in-all, RECO can use N_presamples coming from both MC and data 
RAW for positioning both M0 window and M2 staring time phase. 
I believe unpacker always correcty took pre-samples both from MC 
and data.

Kevin, I'm not sure where from comes 10 pre-samples you quote? I remeber 
unpacker uses 10 TS as a full Digi size, as it's hardcoded in DataFormat
http://cmslxr.fnal.gov/source/DataFormats/HcalDigi/interface/HBHEDataFrame.h#0054

**Comment by slava77** [comment]
@kpedro88 
please follow up on the code review.
It looks like the changes in M0 are expected and probably no other changes are necessary.

**Comment by kpedro88** [comment]
@abdoulline There was a problem specifically with the new packer which was not noticed until this PR. The new packer is supposed to work for any QIE that uses uHTR, which includes QIE8 as well as QIE10/11. However, the QIE8 dataformat is different than QIE10/11, and the presample value for QIE8 was not packed correctly; instead a default value of 10 was used. This meant the unpacked digis also got the value of 10. I fixed this in ec08635309798b575434550d117683d0d1810b1e after discussion with @lihux25 and @jmmans.

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
a new difference popped up in testing of the last changes:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_8_1_X_2016-10-06-1100+16057/16250/validateJR/all_OldVSNew_TTbar13TeV2017HCALdevwf10624p0/all_OldVSNew_TTbar13TeV2017HCALdevwf10624p0c_HcalUnpackerReport_hcalDigis__RECO_obj_unmappedDigis.png

HcalUnpackerReport for hcalDigis has a different size of unmappedDigis

The new test is done with a new GT:
https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/81X_upgrade2017_HCALdev_v2/81X_upgrade2017_HCALdev_Candidate_2016_10_05_11_20_17

Which part of the packer/unpacker changed in this PR leads to this change in the unmappedDigis set?

**Comment by abdoulline** [comment]
~ "The morning is wiser than the evening". 
Now I see... the new unpacked default _was_ incorrectly set to default 10.

Concerning unpapped digis (why do we have any in 2017dev workflow at all once we use consistent 2017dev emap?),  Hongxuan @lihux25 may want to comment  as the new unpacker developer ?

**Comment by abdoulline** [comment]
Concerning  maxZeros  change when N_presamples are fixed - Halil  @dertexaner   
might want to comment. 
This variable reflects the number of Digis with ADC zeros above some limit
http://cmslxr.fnal.gov/source/RecoMET/METProducers/src/HcalNoiseInfoProducer.cc#0411

Frankly, I don't see a simple ("linear") relation between wrong N_presamples in new MC packer and maxZeros (based on checking each ADC count in Digis). But there might be some (impact).

**Comment by slava77** [comment]
signoff of this PR is pending explanations in observed changes
- maxZeros and other noise variables are probably connected to eraw changes
- unmapped digis change is also unexpected, apparently

**Comment by dertexaner** [comment]
RechitR45 and HPD noise filters rely on HBHE eraw (M0), so any change in M0 is likely to affect these filters as well.
MaxZeros (ADC0) filter is no longer in use since mid-2015. But it also has some minimum RBX energy threshold (10 GeV, probably calculated with M0 as well) so it is also plausible that maxZeros distributions get affected by changes to M0 since this could change the number of qualifying RBXs.

**Comment by kpedro88** [comment]
@dertexaner thanks for the info. According to @igv4321 the presample info is used for method 0 (I think), so seems to be a plausible explanation. I will try to check the eraw values explicitly to see if this is reflected. (They're currently not tracked in validation/DQM, which should probably change...)

I'll also try to figure out what's up with the unmapped digis; a change there seems very odd.

**Comment by kpedro88** [comment]
Updates:
1) When I check out a clean IB `CMSSW_8_1_X_2016-09-29-2300` (current basis of this PR) and run 10624.0, `eraw()` for HB RecHits is 0. So I think this explains the changes we see in the noise plots and demonstrates that restoring the proper presample value is indeed a bug fix. (Checking `eraw()` with this PR shows the expected distribution.)
2) I added printouts everywhere `report.countUnmappedDigi(eid)` is called, in a clean IB `CMSSW_8_1_X_2016-10-07-2300` and in the same IB plus this PR. Somehow the list of unmapped QIE8 UTCA electronics IDs does change with the PR included. I really don't know why this would be; the only change for QIE8 should be properly packing the presample value. @lihux25 @jmmans can you comment on this at all?

**Comment by slava77** [comment]
On 10/7/16 8:39 PM, Kevin Pedro wrote:
> Updates:
> 1) When I check out a clean IB |CMSSW_8_1_X_2016-09-29-2300| (current
> basis of this PR) and run 10624.0, |eraw()| for HB RecHits is 0. So I
> think this explains the changes we see in the noise plots and
> demonstrates that restoring the proper presample value is indeed a bug
> fix. (Checking |eraw()| with this PR shows the expected distribution.)

Great. This is an obvious improvement then.

File: EventFilter/HcalRawToDigi/plugins/PackerHelp.h (hunk 1 above)
Comment (slava77) [inline EventFilter/HcalRawToDigi/plugins/PackerHelp.h]: I think this is the reason for the change in unmapped digis.

**Comment by slava77** [comment]
Just dumping the unmapped raw IDs (dumping for 2K events from my pion sample):
- the same set of IDs is present in both cases (144 kinds). There are more unmapped IDs with this PR.
- all appear to be uTCA, at least for one event that I all are 0x40XXXXXX
    - fiberChanId is 3
    - fiberIndex is 31 (0x1F)

... based on finalizeHeadTail method, it looks like these are padding headers 0xD07F  (looks like this is recognized as flavor 5, channel 3 fiber 31).
If I'm not mistaken, the changes here https://github.com/cms-sw/cmssw/pull/16057/files#diff-6929ef99c19e1aac98c2c67529f73efeR519 affect the length of uhtr data and by that the padding size.
So, we get a different number of "unmapped" digis

File: EventFilter/HcalRawToDigi/plugins/PackerHelp.h (hunk 1 above)
Comment (kpedro88) [inline EventFilter/HcalRawToDigi/plugins/PackerHelp.h]: This is only called for QIE11, though, and the unmapped digis are all QIE8.

**Comment by kpedro88** [comment]
@slava77 ah, I was not considering the padding size. You are probably right; including the QIE11 flag word would change the overall size of the raw data stream.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/16057 f09f481
- changes are in line with the description and the follow up review. As mentioned above, the main effect is the fix of presample value for QIE8, which recovers correct values for M0 energy reco in HB (from 0)
- jenkins tests pass and comparisons with baseline show differences in noise flags, which depend on eraw. 
    - There is also a change in the count of unmapped digis, which apparently come out from (still unpacked) empty padding headers. These should eventually be detected and ignored.

**Comment by kpedro88** [comment]
@mdhildreth @civanch please sign again

**Comment by lihux25** [comment]
@slava77 and @kpedro88 thanks for catching the unmapped IDs. We do need padding headers when the word is not 64 bit for QIE8. After talking to Jeremey, he suggested a better way to assign this kind of header. I do plan to have a followup update once this PR is merged.

You are also right that the unmapped headers have no physical impact for now (would be better after following Jeremey's suggest in a later PR).

**Comment by civanch** [comment]
+1

**Comment by kpedro88** [comment]
@davidlange6 can this be considered for pre13?

**Pull request merged**